File: src/bot.js
const { client, xml } = require('@xmpp/client');
const debug = require('@xmpp/debug');
const crypto = require('crypto');
const fs = require('fs-extra');
const path = require('path');
const DB = require('./database');
const config = require('./config');
const PresencePlugin = require('./plugins/presence');
const FileTransferPlugin = require('./plugins/file-transfer');
const CommandsPlugin = require('./plugins/commands');

class OBBFastBot {
    constructor(jid, password, destDir) {
        this.jid = jid;
        this.destDir = destDir;
        this.baseUrl = config.BASE_URL;
        this.pendingFiles = {};

        this.xmpp = client({
            service: `xmpp://${config.XMPP_HOST}:${config.XMPP_PORT}`,
            domain: jid.split('@')[1],
            resource: config.XMPP_RESOURCE,
            username: jid.split('@')[0],
            password: password,
        });

        // debug(this.xmpp); // Uncomment for full XML debugging

        this.db = new DB();
        this.migrateJsonToDb();
        this.migrateFilenames();
        this.cleanupLeftoverParts();

        setInterval(() => this.cleanupPendingFiles(), 60000);

        this.presence = new PresencePlugin(this);
        this.fileTransfer = new FileTransferPlugin(this);
        this.commands = new CommandsPlugin(this);

        this.xmpp.on('error', err => {
            console.error('XMPP Error:', err.message);
        });

        this.xmpp.on('status', status => {
            console.log('XMPP Status:', status);
        });

        this.xmpp.on('stanza', async stanza => {
            if (stanza.is('iq') && stanza.getChild('ping', 'urn:xmpp:ping')) {
                this.handlePing(stanza);
            }
        });
    }

    handlePing(iq) {
        if (iq.attrs.type === 'error' || iq.attrs.type === 'result') return;
        console.log(`PING RECV from ${iq.attrs.from}`);
        const reply = xml('iq', {
            to: iq.attrs.from,
            id: iq.attrs.id,
            type: 'result'
        });
        this.xmpp.send(reply);
        console.log(`PONG SENT to ${iq.attrs.from}`);
    }

    async start() {
        await this.xmpp.start();
    }

    sendMessage(to, body) {
        const message = xml(
            'message',
            { to, type: 'chat' },
            xml('body', {}, body)
        );
        this.xmpp.send(message);
    }

    cleanupPendingFiles() {
        const now = Date.now();
        for (const sid in this.pendingFiles) {
            const info = this.pendingFiles[sid];
            if (info.timestamp && now - info.timestamp > 60000) {
                console.log(`CLEANUP: Removing pending item sid=${sid}`);
                if (info.task && typeof info.task.cancel === 'function') {
                    info.task.cancel();
                }
                delete this.pendingFiles[sid];
            }
        }
    }

    cleanupLeftoverParts() {
        console.log("START: Cleanup leftover .part files");
        let count = 0;
        const walk = (dir) => {
            const files = fs.readdirSync(dir);
            for (const file of files) {
                const fullPath = path.join(dir, file);
                if (fs.statSync(fullPath).isDirectory()) {
                    walk(fullPath);
                } else if (file.endsWith('.part')) {
                    fs.removeSync(fullPath);
                    count++;
                }
            }
        };
        if (fs.existsSync(this.destDir)) {
            walk(this.destDir);
        }
        if (count > 0) console.log(`FINISH: Removed ${count} .part files during startup`);
    }

    migrateJsonToDb() {
        const whitelistFile = config.WHITELIST_FILE;
        try {
            if (fs.existsSync(whitelistFile)) {
                if (fs.lstatSync(whitelistFile).isFile()) {
                    const data = fs.readJsonSync(whitelistFile);
                    for (const entry of data) {
                        this.db.addToWhitelist(entry);
                    }
                    console.log(`MIGRATED ${data.length} entries from ${whitelistFile} to database`);
                    fs.removeSync(whitelistFile);
                } else if (fs.lstatSync(whitelistFile).isDirectory()) {
                    fs.removeSync(whitelistFile);
                }
            }
        } catch (e) {
            console.error(`MIGRATION ERROR: ${e.message}`);
        }
    }

    isAllowed(jid) {
        if (!jid) return false;
        const bareJid = (typeof jid === 'string' ? jid : jid.toString()).split('/')[0].toLowerCase();
        const domain = bareJid.split('@')[1];

        if (config.ADMIN_JID && bareJid === config.ADMIN_JID.toLowerCase()) return true;

        const blacklist = this.db.getBlacklist();
        if (blacklist.has(bareJid) || (domain && blacklist.has(domain))) return false;

        const whitelist = this.db.getWhitelist();
        if (whitelist.has('*')) return true;
        return whitelist.has(bareJid) || (domain && whitelist.has(domain));
    }

    migrateFilenames() {
        console.log("START: Filename migration (spaces to underscores)");
        let count = 0;
        const walk = (dir) => {
            const files = fs.readdirSync(dir);
            for (const file of files) {
                const fullPath = path.join(dir, file);
                if (fs.statSync(fullPath).isDirectory()) {
                    walk(fullPath);
                } else if (file.includes(' ')) {
                    const newPath = path.join(dir, file.replace(/ /g, '_'));
                    fs.renameSync(fullPath, newPath);
                    count++;
                }
            }
        };
        if (fs.existsSync(this.destDir)) {
            walk(this.destDir);
        }
        if (count > 0) console.log(`FINISH: Renamed ${count} files during migration`);
    }

    getUserInfo(jid) {
        const bareJid = (typeof jid === 'string' ? jid : jid.toString()).split('/')[0].toLowerCase();
        let userHash = this.db.getUserFolder(bareJid);
        let isNew = false;
        if (!userHash) {
            userHash = crypto.createHash('md5').update(bareJid).digest('hex');
            this.db.setUserFolder(bareJid, userHash);
            isNew = true;
        }
        const userDir = path.join(this.destDir, userHash);
        if (!fs.existsSync(userDir)) {
            fs.ensureDirSync(userDir);
            isNew = true;
        }
        if (isNew) {
            if (config.ADMIN_JID && (config.ADMIN_NOTIFY_LEVEL === 'all' || config.ADMIN_NOTIFY_LEVEL === 'registrations')) {
                this.sendMessage(config.ADMIN_JID, `🆕 Новый пользователь: ${bareJid} (${userHash})`);
            }
        }
        return { userDir, userHash };
    }

    getHelpText(isAdmin = false, userHash = null) {
        let text = (
            "команды:\n" +
            "ls - список файлов и каталогов в папке пользователя.\n" +
            "ls <-s|-l>, lss, lsl - список файлов (-s: размер, -l: подробно). Пример: ls -l\n" +
            "mkdir <путь> - создать директорию.\n" +
            "rmdir <номер|путь> - удалить пустую директорию.\n" +
            "mv <номер|путь> <номер|путь> - переместить/переименовать.\n" +
            "rm <номер>[,<номер>],.. - удаление файлов по номеру или rm * - для удаления всех файлов.\n" +
            "link <номер>[,<номер>],.. - получение ссылок на файлы или lnk * - для всех файлов.\n" +
            "priv - сделать архив приватным (создать index.html).\n" +
            "pub - сделать архив публичным (удалить index.html).\n" +
            "album - включить режим галереи (создать index.php).\n" +
            "ping - проверить доступность бота.\n" +
            "help или ? - список команд."
        );
        if (userHash) {
            text += `\n\n📂 Ваш архив: ${this.baseUrl}/${userHash}/`;
            text += "\nЧтобы запретить просмотр списка файлов через браузер, используйте команду priv.";
        }
        if (isAdmin) {
            text += (
                "\n\n🔧 Админ-команды:\n" +
                "add <jid|domain|*> - разрешить доступ.\n" +
                "del <jid|domain|*> - запретить доступ.\n" +
                "block <jid|domain> - в чёрный список.\n" +
                "unblock <jid|domain> - убрать из чёрного списка.\n" +
                "list - показать белый и чёрный списки."
            );
        }
        return text;
    }
}

module.exports = OBBFastBot;

File: bot.py
import os
import hashlib
import logging
import asyncio

from slixmpp import ClientXMPP
from slixmpp.xmlstream import ET

from config import (
    ADMIN_JID, VERSION, APP_NAME, BASE_URL
)
from database import Database
from plugins.presence import PresencePlugin
from plugins.commands import CommandsPlugin
from plugins.file_transfer import FileTransferPlugin

class OBBFastBot(ClientXMPP):
    def __init__(self, jid, password, dest_dir):
        super().__init__(jid, password)
        self.dest_dir = dest_dir
        self.base_url = BASE_URL
        self.pending_files = {}
        asyncio.create_task(self.cleanup_pending_files())

        self.db = Database()
        self.migrate_json_to_db()
        self.migrate_filenames()
        self.cleanup_leftover_parts()

        # Initialize core features via plugins
        self.register_plugin('xep_0030')
        self.register_plugin('xep_0047')
        self['xep_0047'].auto_accept = True
        self['xep_0047'].block_size = 32768
        self['xep_0047'].max_block_size = 65536
        self.register_plugin('xep_0199')
        self['xep_0199'].send_keepalive = True
        self['xep_0199'].interval = 60
        self.register_plugin('xep_0092')
        self['xep_0092'].software_name = APP_NAME

        import slixmpp
        import platform
        self['xep_0092'].version = f"{VERSION} on Python {platform.python_version()} + slixmpp {slixmpp.__version__}"

        # Setup custom Ping handler
        from slixmpp.xmlstream import matcher, handler
        self.register_handler(
            handler.Callback('Ping', matcher.MatchXPath('{jabber:client}iq/{urn:xmpp:ping}ping'), self.handle_ping)
        )

        # Load logical modules (plugins)
        self.presence = PresencePlugin(self)
        self.file_transfer = FileTransferPlugin(self)
        self.commands = CommandsPlugin(self)


    def handle_ping(self, iq):
        if iq['type'] in ('error', 'result'): return
        logging.info(f"PING RECV from {iq['from']}")
        reply = iq.reply()
        reply.append(ET.Element('{urn:xmpp:ping}ping'))
        reply.send()
        logging.info(f"PONG SENT to {iq['from']}")

    def make_iq_set(self, ito=None, ifrom=None):
        return self.make_iq(itype='set', ito=ito, ifrom=ifrom)

    async def cleanup_pending_files(self):
        while True:
            try:
                await asyncio.sleep(60)
                now = asyncio.get_event_loop().time()
                to_delete = []
                for sid, info in list(self.pending_files.items()):
                    if isinstance(info, dict):
                        # Timeout for inactive transfers (1 minute)
                        if now - info.get('timestamp', now) > 60:
                            to_delete.append(sid)
                            # Cancel associated task if exists
                            task_key = f"task_{sid}"
                            if task_key in self.pending_files:
                                task = self.pending_files[task_key]
                                if isinstance(task, asyncio.Task) and not task.done():
                                    logging.info(f"CLEANUP: Cancelling inactive task {task_key}")
                                    task.cancel()
                    elif isinstance(info, asyncio.Task):
                        if info.done():
                            to_delete.append(sid)

                for sid in to_delete:
                    if sid in self.pending_files:
                        logging.info(f"CLEANUP: Removing pending item sid={sid}")
                        del self.pending_files[sid]
            except Exception as e:
                logging.error(f"CLEANUP ERROR: {e}")

    def cleanup_leftover_parts(self):
        logging.info("START: Cleanup leftover .part files")
        count = 0
        for root, dirs, files in os.walk(self.dest_dir):
            for f in files:
                if f.endswith('.part'):
                    path = os.path.join(root, f)
                    try:
                        os.remove(path)
                        count += 1
                    except Exception as e:
                        logging.error(f"CLEANUP PART ERROR for {path}: {e}")
        if count > 0:
            logging.info(f"FINISH: Removed {count} .part files during startup")

    def migrate_json_to_db(self):
        whitelist_file = os.getenv('WHITELIST_FILE', 'whitelist.json')
        try:
            if os.path.exists(whitelist_file):
                if os.path.isfile(whitelist_file):
                    import json
                    with open(whitelist_file, 'r') as f:
                        data = json.load(f)
                        for entry in data:
                            self.db.add_to_whitelist(entry)
                    logging.info(f"MIGRATED {len(data)} entries from {whitelist_file} to database")
                    os.remove(whitelist_file)
                elif os.path.isdir(whitelist_file):
                    os.rmdir(whitelist_file)
        except Exception as e:
            logging.error(f"MIGRATION ERROR: {e}")

    def is_allowed(self, jid):
        bare_jid = jid.bare.lower()
        if ADMIN_JID and bare_jid == ADMIN_JID.lower(): return True
        blacklist = self.db.get_blacklist()
        if bare_jid in blacklist or jid.domain.lower() in blacklist: return False
        whitelist = self.db.get_whitelist()
        if '*' in whitelist: return True
        return bare_jid in whitelist or jid.domain.lower() in whitelist

    def migrate_filenames(self):
        logging.info("START: Filename migration (spaces to underscores)")
        count = 0
        for root, dirs, files in os.walk(self.dest_dir):
            for f in files:
                if ' ' in f:
                    old_path = os.path.join(root, f)
                    new_path = os.path.join(root, f.replace(' ', '_'))
                    try:
                        os.rename(old_path, new_path)
                        count += 1
                    except Exception as e:
                        logging.error(f"MIGRATE ERROR for {old_path}: {e}")
        if count > 0:
            logging.info(f"FINISH: Renamed {count} files during migration")

    def get_user_info(self, jid):
        bare_jid = jid.bare.lower()
        user_hash = self.db.get_user_folder(bare_jid)
        is_new = False
        if not user_hash:
            user_hash = hashlib.md5(bare_jid.encode()).hexdigest()
            self.db.set_user_folder(bare_jid, user_hash)
            is_new = True
        user_dir = os.path.join(self.dest_dir, user_hash)
        if not os.path.exists(user_dir): os.makedirs(user_dir); is_new = True
        if is_new:
            from config import ADMIN_NOTIFY_LEVEL
            if ADMIN_JID and ADMIN_NOTIFY_LEVEL in ('all', 'registrations'):
                self.send_message(mto=ADMIN_JID, mbody=f"🆕 Новый пользователь: {bare_jid} ({user_hash})", mtype='chat')
        return user_dir, user_hash

    def get_help_text(self, is_admin=False, user_hash=None):
        text = (
            "команды:\n"
            "ls [фильтр] - список файлов и каталогов в папке пользователя.\n"
            "ls <-s|-l> [фильтр], lss [фильтр], lsl [фильтр] - список файлов (-s: размер, -l: подробно). Примеры: ls -l, lss 1, ls *.jpg\n"
            "mkdir <путь> - создать директорию.\n"
            "rmdir <номер|путь> - удалить пустую директорию.\n"
            "mv <номер|путь> <номер|путь> - переместить/переименовать.\n"
            "rm <номер>[,<номер>],.. - удаление файлов по номеру или rm * - для удаления всех файлов.\n"
            "link <номер>[,<номер>],.. - получение ссылок на файлы или lnk * - для всех файлов.\n"
            "priv - сделать архив приватным (создать index.html).\n"
            "pub - сделать архив публичным (удалить index.html).\n"
            "album - включить режим галереи (создать index.php).\n"
            "ping - проверить доступность бота.\n"
            "help или ? - список команд."
        )
        if user_hash:
            text += f"\n\n📂 Ваш архив: {self.base_url}/{user_hash}/"
            text += "\nЧтобы запретить просмотр списка файлов через браузер, используйте команду priv."
        if is_admin:
            text += (
                "\n\n🔧 Админ-команды:\n"
                "add <jid|domain|*> - разрешить доступ.\n"
                "del <jid|domain|*> - запретить доступ.\n"
                "block <jid|domain> - в чёрный список.\n"
                "unblock <jid|domain> - убрать из чёрного списка.\n"
                "list - показать белый и чёрный списки."
            )
        return text
